Applications/Spire/Source/Ui/FocusObserver.cpp

```diff
@@ -32,6 +32,9 @@ struct FocusObserver::FocusEventFilter : QObject {
   }
 
   void on_focus_changed(QWidget* old, QWidget* now) {
+    if(!now) {
```

Comment:
Is it always the case that if no widget is focused, then the focus isn't considered updated?

Reply:
Good point. The fix is not correct, and I make another fix.
In the new fix, my thinking is when the current focused widget "now" is not m_widget and not the descendant of m_widget, the last focused widget "old" needs to be checked. If the "old" is m_widget or the descendant of m_widget, the focus state is set to `NONE`.

Note that there are two different functions used to check if a widget is another widget's ancestor. Spire::is_ancestor is used to check "now", while QWidget::isAncestorOf is used to check "old". The widget and its ancestor can be in different windows for Spire::is_ancestor, but the widget and its ancestor must be in a same window for QWidget::isAncestorOf.

Suppose there are two OverlayPanels A and B, and they are in different windows.  A is the parent of B, and A is shown. If B is opened, and gets activated and focus, using Spire::is_ancestor to check "now" can get the result that A is B's ancestor and the focus state of A is `FocusIn`, so the focused B doesn't make A closed. When B is closed and loses focus, using QWidget::isAncestorOf to check "old" can get the result that A isn't B's ancestor and the focus state of A doesn't update ('FocusIn'), so the closed B doesn't make A closed.

In this case, A is the OverlayPanel of the OpenFilterPanel and B is the OverlayPanel of the DropDownList inside the ComboBox.